### PR TITLE
Add gather op end-to-end [#754]

### PR DESCRIPTION
**Pull request by ddilbazTT**
Gather is lowered into ttir::EmbeddingOp if possible. ttir::reshapeOp is added to sliceSizes before embeddingOp if needed. Metal currently only supports embedding, not gather.

### CODE REVIEW

**Comment by ddilbazTT** [comment]
Added a TTIR Pass based on [the latest passes refactoring](https://github.com/tenstorrent/tt-mlir/pull/904) Currently, `TTIRGatherPatternMatch` is added to `lib/Dialect/TTIR/Transforms/Transforms.cpp`

File: lib/Dialect/TTIR/Transforms/Transforms.cpp [hunk 1]
```diff
@@ -151,4 +152,123 @@ class TTIRSlidingWindow2dFixShapes
...
+                             collapsedSliceDims.end()) !=
+            std::vector<int64_t>{0}) {
+      return rewriter.notifyMatchFailure(op,
+                                         "Did not satisfy collapsedSliceDims");
+    }
+    if (shape.size() == startIndicesType.getShape().size()) {
+      if (startIndicesType.getShape()[shape.size() - 1] == 1) {
+        reduceWeightTensorDim = true;
+      } else {
+        return rewriter.notifyMatchFailure(op,
```
Comment (uazizTT) [inline lib/Dialect/TTIR/Transforms/Transforms.cpp:197]: Can you please create a function _checkBasicLegality_ and move all the legality checks there?

File: include/ttmlir/Dialect/TTIR/Transforms/Passes.td [hunk 2]
```diff
@@ -134,6 +134,14 @@ def TTIROptimizer: Pass<"ttir-optimizer", "::mlir::ModuleOp"> {
   ];
 }
 
+def TTIRGatherPatternMatch : Pass<"convert-ttir-to-ttir", "::mlir::ModuleOp"> {
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTIR/Transforms/Passes.td:137]: Is the intent of this pass to convert wide TTIR to narrow TTIR?

fyi @sdjordjevicTT

Comment (sdjordjevicTT) [reply include/ttmlir/Dialect/TTIR/Transforms/Passes.td:137]: Please sync with @jserbedzijaTT regarding this pass

Comment (ddilbazTT) [reply include/ttmlir/Dialect/TTIR/Transforms/Passes.td:137]: There is no gather support in metal, but there is embedding support. We decided to keep StableHLO to TTIR layer thin, and then add a pass to convert ttir::gatherOp to ttir::embeddingoP

Comment (sdjordjevicTT) [reply include/ttmlir/Dialect/TTIR/Transforms/Passes.td:111]: Hi @ddilbazTT, @jserbedzijaTT just merged the initial decomposition pass:
https://github.com/tenstorrent/tt-mlir/pull/969

You can reuse it for your scenario.

**Pull request closed without merging**